gson/src/main/java/com/google/gson/internal/bind/ReflectiveTypeAdapterFactory.java

```diff
@@ -226,34 +263,75 @@ private Map<String, BoundField> getBoundFields(Gson context, TypeToken<?> type,
 
   static abstract class BoundField {
     final String name;
+    /** Name of the underlying field */
+    final String componentName;
     final boolean serialized;
     final boolean deserialized;
 
-    protected BoundField(String name, boolean serialized, boolean deserialized) {
+    protected BoundField(String name, String componentName, boolean serialized, boolean deserialized) {
       this.name = name;
+      this.componentName = componentName;
       this.serialized = serialized;
       this.deserialized = deserialized;
     }
-    abstract void write(JsonWriter writer, Object value) throws IOException, IllegalAccessException;
-    abstract void read(JsonReader reader, Object value) throws IOException, IllegalAccessException;
+
+    abstract void write(JsonWriter writer, Object value) throws IOException, ReflectiveOperationException;
+
+    /** Read the value into an array of values, used to construct records */
+    abstract void readIntoArray(JsonReader reader, int index, Object[] values) throws IOException
+
+    /** Read the value from the reader, and set it on the corresponding field on value via reflection */
+    abstract void readIntoField(JsonReader reader, Object value) throws IOException, IllegalAccessException;
   }
 
-  public static final class Adapter<T> extends TypeAdapter<T> {
-    private final ObjectConstructor<T> constructor;
-    private final Map<String, BoundField> boundFields;
+  /**
+   * Base class for Adapters produced by this factory.
+   *
+   * <p>The {@link RecordAdapter} is a special case to handle records for JVMs that support it, for
+   * all other types we use the {@link FieldReflectionAdapter}. This class encapuslates the common
+   * logic for serialization and deserialization. During deserialization, we construct an
+   * intermediary ResultHolder R, which we append values to. After the object has been read in
```

Comment:
Should this be `I` to match the parameter name?
```suggestion
   * intermediary result holder I, which we append values to. After the object has been read in
```

Reply:
I have renamed this to accumulator since I think that name fits better. We accumulate values into the accumulator, then finalize it to create an instance.